R/sim_linelist.R

```diff
@@ -0,0 +1,266 @@
+                         prob_infect,
+                         onset_to_hosp,
+                         onset_to_death,
+                         onset_to_recovery = NA,
+                         hosp_risk = 0.2,
+                         hosp_death_risk = 0.5,
+                         non_hosp_death_risk = 0.05,
+                         outbreak_start_date = as.Date("2023-01-01"),
+                         anonymise = FALSE,
+                         outbreak_size = c(10, 1e4),
+                         population_age = c(1, 90),
+                         case_type_probs = c(
+                           suspected = 0.2,
+                           probable = 0.3,
+                           confirmed = 0.5
+                         ),
+                         config = create_config()) {
+  # check and convert distribution to func if needed before .check_sim_input()
+  funcs <- as_function(
+    list(
+      contact_distribution = contact_distribution,
+      infect_period = infect_period,
+      onset_to_hosp = onset_to_hosp,
+      onset_to_death = onset_to_death,
+      onset_to_recovery = onset_to_recovery
+    )
+  )
+  contact_distribution <- funcs$contact_distribution
+  infect_period <- funcs$infect_period
+  onset_to_hosp <- funcs$onset_to_hosp
+  onset_to_death <- funcs$onset_to_death
+  onset_to_recovery <- funcs$onset_to_recovery
+
+  .check_sim_input(
+    sim_type = "linelist",
+    contact_distribution = contact_distribution,
+    infect_period = infect_period,
+    prob_infect = prob_infect,
+    outbreak_start_date = outbreak_start_date,
+    outbreak_size = outbreak_size,
+    onset_to_hosp = onset_to_hosp,
+    onset_to_death = onset_to_death,
+    onset_to_recovery = onset_to_recovery,
+    anonymise = anonymise,
+    case_type_probs = case_type_probs,
+    hosp_risk = hosp_risk,
+    hosp_death_risk = hosp_death_risk,
+    non_hosp_death_risk = non_hosp_death_risk,
+    population_age = population_age
+  )
+  .cross_check_sim_input(
+    onset_to_hosp = onset_to_hosp,
+    onset_to_death = onset_to_death,
+    hosp_risk = hosp_risk,
+    hosp_death_risk = hosp_death_risk,
+    non_hosp_death_risk = non_hosp_death_risk
+  )
+
+  if (is.data.frame(population_age)) {
+    population_age <- .check_age_df(population_age)
+    age_range <- c(
+      lower = min(population_age[, "min_age"]),
+      upper = max(population_age[, "max_age"])
+    )
+  } else {
+    population_age <- sort(population_age)
+    names(population_age) <- c("lower", "upper")
+    age_range <- population_age
+  }
+  if (is.data.frame(hosp_risk)) {
+    hosp_risk <- .check_risk_df(
+      hosp_risk,
+      age_range = age_range
+    )
+  }
+  if (is.data.frame(hosp_death_risk)) {
+    hosp_death_risk <- .check_risk_df(
+      hosp_death_risk,
+      age_range = age_range
+    )
+  }
+  if (is.data.frame(non_hosp_death_risk)) {
+    non_hosp_death_risk <- .check_risk_df(
+      non_hosp_death_risk,
+      age_range = age_range
+    )
+  }
```

Comment:
I agree that this looks like a code smell. Thanks for the suggestions on how to refactor this chunk. I've made some changes from these suggestions on a branch called `refactor_df_checks`. 

However, I'm not sure that I prefer the new structure. Although ideally we'd be able to remove the chain of `if` statements all performing the same check, when implemented in it's own function the risks need to be bundled into a list, the new function adds more code to the package, function documentation and unit tests (yet to write) are required, and then the checked objects require unpacking in the `sim_*()` functions. 

_Note_ the `.check_risk_df()` does some `<data.frame>` formatting and so is not called purely for side-effects of warnings or errors. Therefore returning the objects and reassigning them to variables in required. 

Please let me know your thoughts on this refactor and which you prefer, or if the refactor I've implemented is different from what you had in mind. I will leave the `refactor_df_checks` branch open on the package so if we decide to make this change it can be merged after the v0.3.0 release.